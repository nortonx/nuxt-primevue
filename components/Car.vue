<template>
  <div class="car-item border-1 border-round-sm p-2 w-2 flex flex-column justify-content-center align-items-center m-3">
    <div class="action-buttons flex flex-wrap justify-content-center align-items-center m-1">
      <Button
        icon="pi pi-pencil"
        @click="editCar"
        :label="editLabel"
        class="mb-1"
        severity="info"
      ></Button>
      <Button
        icon="pi pi-trash"
        @click="deleteCar"
        label="Delete"
        class="mb-1 ml-1"
        severity="danger"
      ></Button>
    </div>
    <img 
      src="~/assets/car.svg"
      :alt="props.carModel"
      class="car-image p-1"
    />
    <div class="car-info">
      <editable-field label="Model" :value="props.carModel" v-model="props.carModel" class="car-model" @toggle-click="toggleClick"></editable-field>
      <editable-field label="Plate" :value="props.carPlate" v-model="props.carPlate" @toggle-click="toggleClick"></editable-field>
      <editable-field label="Color" :value="props.carColor" v-model="props.carColor" @toggle-click="toggleClick"></editable-field>
      <div class="car-owner"><span class="car-property font-bold">Owner</span>: {{ props.carOwner }}</div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import Button from "primevue/button"
import type { Car } from "@/types/car.type"

const API_URL = import.meta.env.VITE_PARKING_API_URL as string
const props = defineProps<{
  id: number,
  carModel: string,
  carPlate: string,
  carColor: string,
  carOwner: string
}>()

const clicked = ref(false)
const editLabel = ref("Edit")

function toggleClick() {
  clicked.value = !clicked.value
  editLabel.value = clicked.value ? "Save" : "Edit"
}

async function editCar() {
  console.log("Editing car with plate:", props.carPlate, props.carModel)
  const { data, error, status } = await useFetch<Car>(`${API_URL}/cars/${props.id}`, {
    method: "PUT",
    body: JSON.stringify({
      model: props.carModel,
      plate: props.carPlate,
      color: props.carColor,
      user: props.carOwner
    })
  })
  console.log("Data from parking-api:", data.value, "error:", error.value, "status:", status.value)
  toggleClick()
}

function deleteCar() {
  console.log("Deleting car with plate:", props.carPlate, props.carModel)
}
</script>

<style scoped>
.car-item {
  min-width: 160px;
  img {
    width: 120px;
    height: 120px;
  }
}
</style>