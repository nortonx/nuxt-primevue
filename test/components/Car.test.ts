import { shallowMount } from "@vue/test-utils";
import Car from "@/components/Car.vue";
import { describe, it, expect, beforeEach } from "vitest";

describe("Car component", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(Car, {
      props: {
        carModel: "Toyota",
        carPlate: "ABC-1234",
        carColor: "black",
        carOwner: "John Doe",
      },
    });
  });

  it("should render correctly and match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render correctly with props and match snapshot", () => {
    expect(wrapper.find(".car-model").text()).toBe(
      `Model: ${wrapper.props().carModel}`,
    );
    expect(wrapper.find(".car-plate").text()).toBe(
      `Plate: ${wrapper.props().carPlate}`,
    );
    expect(wrapper.find(".car-color").text()).toBe(
      `Color: ${wrapper.props().carColor}`,
    );
    expect(wrapper.find(".car-owner").text()).toBe(
      `Owner: ${wrapper.props().carOwner}`,
    );

    expect(wrapper.html()).toMatchSnapshot();
  });
});
