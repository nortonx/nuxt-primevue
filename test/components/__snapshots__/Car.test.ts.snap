// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Car component > should render correctly and match snapshot 1`] = `
"<div data-v-0a0a5ccd="" class="car-item border-1 border-round-sm p-2 w-2 flex flex-column justify-content-center align-items-center m-3"><img data-v-0a0a5ccd="" src="/assets/car-compact-svgrepo-com.svg" alt="Toyota" class="car-image p-1">
  <div data-v-0a0a5ccd="" class="car-info">
    <div data-v-0a0a5ccd="" class="car-model"><span data-v-0a0a5ccd="" class="car-property font-bold">Model</span>: Toyota</div>
    <div data-v-0a0a5ccd="" class="car-plate"><span data-v-0a0a5ccd="" class="car-property font-bold">Plate</span>: <code data-v-0a0a5ccd="">ABC-1234</code></div>
    <div data-v-0a0a5ccd="" class="car-color"><span data-v-0a0a5ccd="" class="car-property font-bold">Color</span>: <span data-v-0a0a5ccd="" class="car-property-value" style="border: 2px solid black; padding-left: 2px; padding-right: 2px;">black</span></div>
    <div data-v-0a0a5ccd="" class="car-owner"><span data-v-0a0a5ccd="" class="car-property font-bold">Owner</span>: John Doe</div>
  </div>
</div>"
`;

exports[`Car component > should render correctly with props and match snapshot 1`] = `
"<div data-v-0a0a5ccd="" class="car-item border-1 border-round-sm p-2 w-2 flex flex-column justify-content-center align-items-center m-3"><img data-v-0a0a5ccd="" src="/assets/car-compact-svgrepo-com.svg" alt="Toyota" class="car-image p-1">
  <div data-v-0a0a5ccd="" class="car-info">
    <div data-v-0a0a5ccd="" class="car-model"><span data-v-0a0a5ccd="" class="car-property font-bold">Model</span>: Toyota</div>
    <div data-v-0a0a5ccd="" class="car-plate"><span data-v-0a0a5ccd="" class="car-property font-bold">Plate</span>: <code data-v-0a0a5ccd="">ABC-1234</code></div>
    <div data-v-0a0a5ccd="" class="car-color"><span data-v-0a0a5ccd="" class="car-property font-bold">Color</span>: <span data-v-0a0a5ccd="" class="car-property-value" style="border: 2px solid black; padding-left: 2px; padding-right: 2px;">black</span></div>
    <div data-v-0a0a5ccd="" class="car-owner"><span data-v-0a0a5ccd="" class="car-property font-bold">Owner</span>: John Doe</div>
  </div>
</div>"
`;
